### ISSUE TITLE
Prolog crashes when using transaction/1

### ISSUE THREAD

**Issue by EricZinda** [SWI-Prolog/swipl-devel#810]
In development build 8.3.20, threaded, 64 bits on a mac, I can reliably reproduce a crash when using the transaction/1 term in a really complex program.  Here is the predicate that uses the transaction:
~~~
d_udef_q__xhh(VariableArg, RstrArg, BodyArg, QuoteOrEvalArg) :-  arg(Variable, _) = VariableArg, arg(Rstr, _) = RstrArg, arg(Body, _) = BodyArg, arg(Quote, _) = QuoteOrEvalArg,
    (
        getContext(abduce)
        ->
        (
            % Must be a concept (i.e. \+ isInstance(Variable)
            X = abduceNow(conj(Rstr, conj(tryError(\+ isInstance(Variable), context(d_udef_q__xhh, contextNoConcepts, VariableArg)))))

        )
        ;
        (
            % Must be a concept (i.e. \+ isInstance(Variable)), and make sure nothing in here is abducted by using check()
            X = check(conj(Rstr, conj(tryError(\+ isInstance(Variable), context(d_udef_q__xhh, contextNoConcepts, VariableArg)))))
        )
    ),

    transaction((
        resolveTerm(X, Quote),
        % Then run the body
        resolveTerm(Body, Quote)
        )),
    % if the concept is what we needed, we are done
    !.
~~~
If I comment out the transaction/1 term it is fine.  It is going to be really hard to get a repro case you can run since this is a reasonably large body of code.  I do have the crash dump file: 
[crash.txt](https://github.com/SWI-Prolog/swipl-devel/files/6261035/crash.txt)

Is this enough to narrow it down?

**Comment by EricZinda** [comment]
@JanWielemaker I realized you have access to the source.  I created a branch called "Bug810".  It reliably repros with the following sequence:
~~~
?- consult(".../NLTK/Debug.pl").
**** Scenario ****
true.

?- abduceProposition(freeVariables(erg(0, d_the_q__xhh(arg(X9002, var([name(x10), type(reg), pers(3), num(pl)])), arg(erg(1, d_noun__x(arg('rock', word('rock')), arg(X9002, var([name(x10), type(reg), pers(3), num(pl)])))), term), arg(erg(2, d_udef_q__xhh(arg(X9004, var([name(x4), type(reg), pers(3), num(sg)])), arg(conj(erg(3, d_noun__x(arg('dust', word('dust')), arg(X9004, var([name(x4), type(reg), pers(3), num(sg)])))), conj(erg(4, d_on_p_loc__exx(arg(E9006, var([name(e9), tense(untensed)])), arg(X9004, var([name(x4), type(reg), pers(3), num(sg)])), arg(X9002, var([name(x10), type(reg), pers(3), num(pl)])), arg(Quote9011, var([type(quote)])))))), term), arg(erg(5, conj(d_be_v_there__ex(arg(E9007, var([name(e2), index(yes), tense(pres)])), arg(X9004, var([name(x4), type(reg), pers(3), num(sg)])), arg(create, var([type(quote)]))), conj(erg(6, applyFirstPlan([task_be_v_there__ex(arg(E9007, var([name(e2), index(yes), tense(pres)])))], Solution))))), term), arg(Quote9010, var([type(quote)])))), term), arg(Quote9008, var([type(quote)])))), [item(x10, X9002), item(x4, X9004), item(h6, H9005), item(e9, E9006), item(e2, E9007)])).

>>> Crash
~~~

**Comment by JanWielemaker** [comment]
Thanks.   That made the job a lot simpler.   Fixed with 2fc09cd8976c2ddbf3a2cb4cd87877d6743ef1a3

**Comment by EricZinda** [comment]
Thanks @JanWielemaker.  Fix works great1
